src/lib/PerceptronPredictor.cc

```diff
@@ -0,0 +1,159 @@
+  // non-zero bits of the address, and then keep only the btbBits_ bits of the output
+  // keep it in bounds of the prediction table.
+  uint64_t hashedIndex = ((address >> 2) ^ globalHistory_) & ((1 << btbBits_) - 1);
+
+  // Store the global history for correct hashing in update() --
+  // needs to be global history and not the hashed index as hashing looses information at longer
+  // global history lengths
+  btbHistory_[address] = globalHistory_;
+
+  // Retrieve the perceptron from the BTB
+  std::vector<int8_t> perceptron = btb_[hashedIndex].first;
+
+  // Determine direction prediction from perceptron by taking its dot product
+  // with the global history
+  // Starting with the bias weight
+  int64_t Pout = perceptron[globalHistoryLength_];
+  for (int i = 0; i < globalHistoryLength_; i++) {
+    // Get branch direction for ith entry in the globalHistory_
+    bool historyTaken =
+        ((globalHistory_ & (1 << ((globalHistoryLength_ - 1) - i))) != 0);
+    Pout += historyTaken ? perceptron[i] : (0 - perceptron[i]);
+  }
+  bool direction = (Pout >= 0);
+
+  // Retrieve target prediction
+  uint64_t target =
+      (knownOffset != 0) ? address + knownOffset : btb_[hashedIndex].second;
+
+  BranchPrediction prediction = {direction, target};
+
+  // Amend prediction based on branch type
+  if (type == BranchType::Unconditional) {
+    prediction.taken = true;
+  } else if (type == BranchType::Return) {
+    prediction.taken = true;
+    // Return branches can use the RAS if an entry is available
+    if (ras_.size() > 0) {
+      prediction.target = ras_.back();
+      // Record top of RAS used for target prediction
+      rasHistory_[address] = ras_.back();
+      ras_.pop_back();
+    }
+  } else if (type == BranchType::SubroutineCall) {
+    prediction.taken = true;
+    // Subroutine call branches must push their associated return address to RAS
+    if (ras_.size() >= rasSize_) {
+      ras_.pop_front();
+    }
+    ras_.push_back(address + 4);
+    // Record that this address is a branch-and-link instruction
+    rasHistory_[address] = 0;
+  } else if (type == BranchType::Conditional) {
+    if (!prediction.taken) prediction.target = address + 4;
+  }
+  return prediction;
+}
+
+void PerceptronPredictor::update(uint64_t address, bool taken,
+                                 uint64_t targetAddress, BranchType type) {
+  // Work out hash index
+  uint64_t prevGlobalHistory = btbHistory_[address];
+  uint64_t hashedIndex =
+      ((address >> 2) ^ prevGlobalHistory) & ((1 << btbBits_) - 1);
+
+  std::vector<int8_t> perceptron = btb_[hashedIndex].first;
+
+  // Work out the most recent prediction
+  int64_t Pout = perceptron[globalHistoryLength_];
+  for (int i = 0; i < globalHistoryLength_; i++) {
+    bool historyTaken =
+        ((prevGlobalHistory & (1 << ((globalHistoryLength_ - 1) - i))) != 0);
+    Pout += historyTaken ? perceptron[i] : (0 - perceptron[i]);
+  }
+  bool directionPrediction = (Pout >= 0);
```

Comment:
This section of code is identical to that in predict(). You could create a new private function to reduce repeated code

Reply:
I prefer the private function -- in reality I don't think that the dot product is stored between prediction and update (to reduce hardware cost)